streams/src/test/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionSendProcessorSupplierTest.java

```diff
@@ -0,0 +1,302 @@
+        );
+    }
+
+    @Test
+    public void leftJoinShouldPropagateChangeOfFKFromNonNullToNonNullValue() {
+        final MockInternalNewProcessorContext<String, SubscriptionWrapper<String>> context = new MockInternalNewProcessorContext<>();
+        leftJoinProcessor.init(context);
+        context.setRecordMetadata("topic", 0, 0);
+
+        final LeftValue leftRecordValue = new LeftValue("merchant2");
+
+        leftJoinProcessor.process(new Record<>("product1", new Change<>(leftRecordValue, new LeftValue("merchant1")), 0));
+
+        assertThat(context.forwarded().size(), is(2));
+        assertThat(
+            context.forwarded().get(1).record(),
+            is(new Record<>("merchant2", new SubscriptionWrapper<>(hash(leftRecordValue), PROPAGATE_NULL_IF_NO_FK_VAL_AVAILABLE, "product1", 0), 0))
+        );
+    }
+
+    @Test
+    public void leftJoinShouldPropagateChangeOfFKFromNonNullToNullValue() {
+        final MockInternalNewProcessorContext<String, SubscriptionWrapper<String>> context = new MockInternalNewProcessorContext<>();
+        leftJoinProcessor.init(context);
+        context.setRecordMetadata("topic", 0, 0);
+
+        final LeftValue leftRecordValue = new LeftValue(null);
+
+        leftJoinProcessor.process(new Record<>("product1", new Change<>(leftRecordValue, new LeftValue("merchant1")), 0));
+
+        assertThat(context.forwarded().size(), greaterThan(0));
+        assertThat(
+            context.forwarded().get(0).record(),
+            is(new Record<>("merchant1", new SubscriptionWrapper<>(hash(leftRecordValue), DELETE_KEY_AND_PROPAGATE, "product1", 0), 0))
+        );
+    }
+
+    @Test
+    public void leftJoinShouldPropagateChangeFromNullFKToNullFKValue() {
+        final MockInternalNewProcessorContext<String, SubscriptionWrapper<String>> context = new MockInternalNewProcessorContext<>();
+        leftJoinProcessor.init(context);
+        context.setRecordMetadata("topic", 0, 0);
+
+        final LeftValue leftRecordValue = new LeftValue(null);
+
+        leftJoinProcessor.process(new Record<>("product1", new Change<>(leftRecordValue, leftRecordValue), 0));
+
+        assertThat(context.forwarded().size(), is(1));
+        assertThat(
+            context.forwarded().get(0).record(),
+            is(new Record<>(null, new SubscriptionWrapper<>(hash(leftRecordValue), PROPAGATE_NULL_IF_NO_FK_VAL_AVAILABLE, "product1", 0), 0))
+        );
+    }
+
+    @Test
+    public void leftJoinShouldPropagateDeletionOfAPrimaryKey() {
+        final MockInternalNewProcessorContext<String, SubscriptionWrapper<String>> context = new MockInternalNewProcessorContext<>();
+        leftJoinProcessor.init(context);
+        context.setRecordMetadata("topic", 0, 0);
+
+        leftJoinProcessor.process(new Record<>("product1", new Change<>(null, new LeftValue("merchant1")), 0));
+
+        assertThat(context.forwarded().size(), greaterThan(0));
```

Comment:
Didn't test the exact number of records because of the bug KAFKA-16394, this test is to be made stronger after fixing the bug.